google-cloud-sdk/help/man/man1/gcloud.1
.TH "GCLOUD " "1" "" "" ""
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.SH "NAME"
.HP
gcloud \- manage Google Cloud Platform resources and developer workflow
.SH "SYNOPSIS"
.sp
gcloud \fIGROUP\fR | \fICOMMAND\fR [\fB\-\-format\fR \fIFORMAT\fR] [\fB\-\-help\fR] [\fB\-\-log\-http\fR] [\fB\-\-project\fR \fIPROJECT_ID\fR] [\fB\-\-quiet\fR, \fB\-q\fR] [\fB\-\-trace\-token\fR \fITRACE_TOKEN\fR] [\fB\-\-user\-output\-enabled\fR \fIUSER_OUTPUT_ENABLED\fR] [\fB\-\-verbosity\fR \fIVERBOSITY\fR] [\fB\-h\fR] [\fB\-v\fR, \fB\-\-version\fR]
.SH "DESCRIPTION"
.sp
The \fBgcloud\fR CLI manages authentication, local configuration, developer workflow, and interactions with the Google Cloud Platform APIs\&.
.SH "FLAGS"
.HP
\fB\-\-format\fR \fIFORMAT\fR
json,
text,
yaml\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Display detailed help\&.
.RE
.PP
\fB\-\-log\-http\fR
.RS 4
Logs all HTTP server requests and responses to stderr\&.
.RE
.PP
\fB\-\-project\fR \fIPROJECT_ID\fR
.RS 4
The Google Cloud Platform project name to use for this invocation\&. If omitted then the current project is assumed\&.
.RE
.PP
\fB\-\-quiet\fR, \fB\-q\fR
.RS 4
Disable all interactive prompts when running gcloud commands\&. If input is required, defaults will be used, or an error will be raised\&.
.RE
.PP
\fB\-\-trace\-token\fR \fITRACE_TOKEN\fR
.RS 4
Token used to route traces of service requests for investigation of issues\&.
.RE
.PP
\fB\-\-user\-output\-enabled\fR \fIUSER_OUTPUT_ENABLED\fR
.RS 4
Control whether user intended output is printed to the console\&. (true/false)
.RE
.PP
\fB\-\-verbosity\fR \fIVERBOSITY\fR
.RS 4
Override the default verbosity for this command\&. This must be a standard logging verbosity level: [debug, info, warning, error, critical, none] (Default: [warning])\&.
.RE
.PP
\fB\-h\fR
.RS 4
Print a summary help and exit\&.
.RE
.SH "GROUP FLAGS"
.PP
\fB\-v\fR, \fB\-\-version\fR
.RS 4
Print version information\&.
.RE
.SH "GROUPS"
.sp
\fIGROUP\fR is one of the following:
.PP
\fBalpha\fR
.RS 4
\fB(ALPHA)\fR
Alpha versions of gcloud commands\&.
.RE
.PP
\fBauth\fR
.RS 4
Manage oauth2 credentials for the Google Cloud SDK\&.
.RE
.PP
\fBbeta\fR
.RS 4
\fB(BETA)\fR
Beta versions of gcloud commands\&.
.RE
.PP
\fBcomponents\fR
.RS 4
List, install, update, or remove Google Cloud SDK components or packages\&.
.RE
.PP
\fBcompute\fR
.RS 4
Read and manipulate Google Compute Engine resources\&.
.RE
.PP
\fBconfig\fR
.RS 4
View and edit Google Cloud SDK properties\&.
.RE
.PP
\fBdns\fR
.RS 4
Manage your Cloud DNS managed\-zones and record\-sets\&.
.RE
.PP
\fBpreview\fR
.RS 4
Manage Preview CLI command groups\&.
.RE
.PP
\fBsql\fR
.RS 4
Manage Cloud SQL databases\&.
.RE
.PP
\fBtopic\fR
.RS 4
gcloud supplementary help\&.
.RE
.SH "COMMANDS"
.sp
\fICOMMAND\fR is one of the following:
.PP
\fBdocker\fR
.RS 4
Provides the docker CLI access to the Google Container Registry\&.
.RE
.PP
\fBhelp\fR
.RS 4
Prints detailed help messages for the specified commands\&.
.RE
.PP
\fBinfo\fR
.RS 4
Display information about the current gcloud environment\&.
.RE
.PP
\fBinit\fR
.RS 4
Create and initialize a gcloud workspace in the current directory\&.
.RE
.PP
\fBversion\fR
.RS 4
Print version information for Cloud SDK components\&.
.RE
.SH "NOTES"
.sp
This command is in the Google Cloud SDK \fBcore\fR component\&. See installing components if it is not installed\&.


google-cloud-sdk/RELEASE_NOTES
Copyright 2014-2015 Google Inc.
All rights reserved.


Google Cloud SDK - Release Notes

0.9.67 (2015/07/01)
==================
  - GAE components updated to 1.9.23. Please visit
    https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes for details.
  - Added the "gen-config" command to app surface to generate configuration
    files for node.js apps.
  - New `compute instance-groups unmanaged` commands group for Instance Groups
    which are not managed by Instance Group Manager
  - Moved `gcloud beta sql instances connect` to `gcloud beta sql connect`.
  - Smarter polling while waiting for SQL operations to finish, reducing the
    number of requests sent to the server.
  - Linux/Mac install script disables prompts if there is no controlling tty.
  - Linux/Mac install script now properly detects zsh and ksh RC files.
  - Bug fixes.

0.9.66 (2015/06/24)
==================

  - New commands
    - `gcloud alpha compute instances set-machine-type` for managing types
    - `gcloud topic` for supplementary help topics.
  - `gcloud deployment-manager deployments delete` prompt defaults to false.

0.9.65 (2015/06/17)
==================
- Fixes `reset-windows-password` on Windows if `openssl` is present on the path.
- `gcloud compute reset-windows-password` is now available as a GA command.
- Prevent creation of a user that matches the machine name on Windows.
- gsutil now runs properly from not-default drive letters on Windows.
- kubectl updated to 0.18.1.
- `gcloud preview docker` moved to `gcloud docker`.

0.9.64 (2015/06/10)
==================
  - New commands
    - `gcloud beta compute scp` for secure copy.
    - `gcloud alpha genomics` commands (on datasets).
    - `gcloud alpha source list` to lists project repositories.
  - kubectl will no longer be installed by default.
  - Bug fixes for https load balancing.
  - The help pager is $PAGER or less(1) if available, otherwise a builtin pager.
  - Update of gsutil component to version 4.13.

0.9.63 (2015/06/03)
==================
  - 2x faster command load times.
  - Fix to allow `preview app` to use oauth service accounts
    https://code.google.com/p/google-cloud-sdk/issues/detail?id=133

0.9.62 (2015/05/27)
==================
  - Not released.

0.9.61 (2015/05/20)
==================
  - New kubectl component (installed by default)
  - Internal improvements to gcloud preview app deploy; includes changes to
    command output.
  - `gcloud preview projects` moved to `gcloud alpha projects`
  - Tab completion for `gcloud compute ssh`

0.9.60 (2015/05/13)
===================
  - VPN moved to GA.
  - Preemptible VM support.
  - Bug fixes.

0.9.59 (2015/05/06)
===================
  - Bug fixes.

0.9.58 (2015/04/29)
===================
  - gcloud alpha bigquery launch.
  - Better error reporting in Customer Supplied Encryption Key files.
  - Bug fixes.

0.9.57 (2015/04/22)
===================
  - gcloud beta deployment-manager launch.
  - gcloud app new deployment behavior.
    - deploy and run commands no longer take directories, only take a list of
      explicit yaml files.
  - gcloud beta compute reset-windows-password, support for new authentication
    scheme on Windows guest agent versions 3.0.0.0 and later.
  - Bug fixes.

0.9.56 (2015/04/15)
===================
  - Bug fixes.

0.9.55 (2015/04/8)
===================
 - New and improved gcloud dns surface including:
   - Transactions for scriptable editing of record-sets.
   - Importing record-sets from zone files and yaml files.
   - Exporting record-sets to zone files and yaml files.
 - VPN commands surfaced under gcloud compute.
 - Dataflow CLI commands launched under gcloud alpha!

0.9.52 (2015/03/18)
===================
  - Bug Fixes.

0.9.51 (2015/03/11)
===================
  - Bug Fixes.

0.9.50 (2015/03/04)
===================
  - Introducing the `gcloud beta ...` command group
    - Existing gcloud commands will be replicated under this new group, but use
      the beta version of their API instead of the GA version.
    - New command behavior and features that are beta only will first appear
      under this group until the feature gets promoted to general availability.
    - Run `gcloud components update beta` to access the new commands, or it will
      be automatically installed for you on first run attempt.
  - `gcloud beta compute ...` is now available, using the compute 'beta' API.
  - Beta release of Cloud VPN.  See https://cloud.google.com/compute/docs/vpn
    for details.
  - Release of `gcloud compute instances move` and `gcloud compute disks move`.

0.9.49 (2015/02/25)
===================
  - The Google Container Registry is now used by default for deployment of
    Managed VM applications. See here for more information:
    https://cloud.google.com/tools/container-registry/
  - We've increased the speed of the "image push" phase of Managed VM
    deployments.
  - The heavy lifting of retrieving the Go toolchain has been moved into the
    base Go Docker image which dramatically reduces the bandwidth consumed
    locally when running and deploying your images. You must delete your
    Dockerfile to pick up the changes.
  - We've fixed several bugs related to port forwarding in the local
    devappserver.

0.9.48 (2015/02/18)
===================
  - A bug fix for running setup-managed-vms multiple times when using the Google
    Container Registry

0.9.47 (2015/02/11)
===================
  - Bug fixes that allow tab completion to work for components in preview.
  - GCloud now auto installs components based on command execution.

0.9.46 (2015/02/04)
===================
  - We've made significant improvements to the stability and speed of deploying
    Managed VMs by leveraging the new Google Container Registry. See this post
    for details on how to try it out today:
    https://groups.google.com/forum/#!topic/google-appengine/c543NHZGhKM

0.9.45 (2015/01/28)
===================
  - Several bug fixes in the Dockerfile generated for the 'go' runtime. Delete

google-cloud-sdk/lib/googlecloudsdk/gcloud/gcloud.py
# Copyright 2013 Google Inc. All Rights Reserved.

"""gcloud command line tool."""

import time
START_TIME = time.time()

# pylint:disable=g-bad-import-order
# pylint:disable=g-import-not-at-top, We want to get the start time first.
import os
import signal
import sys
import traceback


# Disable stack traces when people kill a command.
def CTRLCHandler(unused_signal, unused_frame):
  """Custom SIGNINT handler.

  Signal handler that doesn't print the stack trace when a command is
  killed by keyboard interupt.
  """
  try:
    log.err.Print('\n\nCommand killed by keyboard interrupt\n')
  except NameError:
    sys.stderr.write('\n\nCommand killed by keyboard interrupt\n')
  # Kill ourselves with SIGINT so our parent can detect that we exited because
  # of a signal. SIG_DFL disables further KeyboardInterrupt exceptions.
  signal.signal(signal.SIGINT, signal.SIG_DFL)
  os.kill(os.getpid(), signal.SIGINT)
  # Just in case the kill failed ...
  sys.exit(1)
signal.signal(signal.SIGINT, CTRLCHandler)


# Enable normal UNIX handling of SIGPIPE to play nice with grep -q, head, etc.
# See https://mail.python.org/pipermail/python-list/2004-June/273297.html and
# http://utcc.utoronto.ca/~cks/space/blog/python/SignalExceptionSurprise
# for more details.
if hasattr(signal, 'SIGPIPE'):
  signal.signal(signal.SIGPIPE, signal.SIG_DFL)


def _SetPriorityCloudSDKPath():
  """Put google-cloud-sdk/lib at the beginning of sys.path.

  Modifying sys.path in this way allows us to always use our bundled versions
  of libraries, even when other versions have been installed. It also allows the
  user to install extra libraries that we cannot bundle (ie PyOpenSSL), and
  gcloud commands can use those libraries.
  """

  def _GetRootContainingGoogle():
    path = __file__
    while True:
      parent, here = os.path.split(path)
      if not here:
        break
      if here == 'googlecloudsdk':
        return parent
      path = parent

  module_root = _GetRootContainingGoogle()

  # check if we're already set
  if (not module_root) or (sys.path and module_root == sys.path[0]):
    return
  sys.path.insert(0, module_root)


def _DoStartupChecks():
  from googlecloudsdk.core.util import platforms
  if not platforms.PythonVersion().IsSupported():
    sys.exit(1)
  if not platforms.Platform.Current().IsSupported():
    sys.exit(1)


_SetPriorityCloudSDKPath()
_DoStartupChecks()


# pylint:disable=g-import-not-at-top, We want the _SetPriorityCloudSDKPath()
# function to be called before we try to import any CloudSDK modules.
from googlecloudsdk.calliope import base
from googlecloudsdk.calliope import cli
from googlecloudsdk.core import config
from googlecloudsdk.core import log
from googlecloudsdk.core import metrics
from googlecloudsdk.core import properties
from googlecloudsdk.core.updater import local_state
from googlecloudsdk.core.updater import update_manager


if not config.Paths().sdk_root:
  # Don't do update checks if there is no install root.
  properties.VALUES.component_manager.disable_update_check.Set(True)


def UpdateCheck():
  try:
    update_manager.UpdateManager(warn=False).PerformUpdateCheck()
  # pylint:disable=broad-except, We never want this to escape, ever. Only
  # messages printed should reach the user.
  except Exception:
    pass


def VersionFunc():
  _cli.Execute(['version'])


def CreateCLI():
  """Generates the gcloud CLI."""
  pkg_root = config.GoogleCloudSDKPackageRoot()
  loader = cli.CLILoader(
      name='gcloud',
      command_root_directory=os.path.join(
          pkg_root, 'gcloud', 'sdktools', 'root'),
      allow_non_existing_modules=True,
      version_func=VersionFunc)

  loader.AddReleaseTrack(base.ReleaseTrack.ALPHA,
                         os.path.join(pkg_root, 'gcloud', 'sdktools', 'alpha'),
                         component='alpha')
  loader.AddReleaseTrack(base.ReleaseTrack.BETA,
                         os.path.join(pkg_root, 'gcloud', 'sdktools', 'beta'),
                         component='beta')

  loader.AddModule('auth', os.path.join(pkg_root, 'gcloud', 'sdktools', 'auth'))
  loader.AddModule('bigquery', os.path.join(pkg_root, 'bigquery', 'commands'))
  loader.AddModule('bigtable', os.path.join(pkg_root, 'bigtable', 'commands'))
  loader.AddModule('components',
                   os.path.join(pkg_root, 'gcloud', 'sdktools', 'components'))
  loader.AddModule('compute', os.path.join(pkg_root, 'compute', 'subcommands'),
                   component='compute')
  loader.AddModule('config',
                   os.path.join(pkg_root, 'gcloud', 'sdktools', 'config'))
  loader.AddModule('container',
                   os.path.join(pkg_root, 'container', 'commands'),
                   component='gcloud')
  loader.AddModule('dataflow',
                   os.path.join(pkg_root, 'dataflow', 'commands'))
  loader.AddModule('deployment_manager',
                   os.path.join(pkg_root, 'deployment_manager', 'commands'))
  loader.AddModule('dns', os.path.join(pkg_root, 'dns', 'commands'),
                   component='dns')
  loader.AddModule('endpoints', os.path.join(pkg_root, 'endpoints', 'commands'))
  loader.AddModule('internal',
                   os.path.join(pkg_root, 'gcloud', 'sdktools', 'internal'))
  loader.AddModule('meta', os.path.join(pkg_root, 'gcloud', 'sdktools', 'meta'))
  loader.AddModule('preview', os.path.join(pkg_root, 'preview', 'commands'),
                   component='preview')
  # Put app and datastore under preview for now.
  loader.AddModule('preview.analysis',
                   os.path.join(pkg_root, 'analysis', 'commands'))
  loader.AddModule('preview.app',
                   os.path.join(pkg_root, 'appengine', 'app_commands'),
                   component='app')
  loader.AddModule('preview.datastore',
                   os.path.join(pkg_root, 'appengine', 'datastore_commands'),
                   component='app')
  loader.AddModule('genomics',
                   os.path.join(pkg_root, 'genomics', 'commands'))
  # TODO(user): Put logging in preview for now.
  loader.AddModule('preview.logging',
                   os.path.join(pkg_root, 'logging', 'commands'))
  loader.AddModule('preview.rolling_updates',
                   os.path.join(
                       pkg_root, 'updater', 'commands', 'rolling_updates'))
  loader.AddModule('projects',
                   os.path.join(pkg_root, 'projects', 'commands'),
                   component='gcloud')
  loader.AddModule('pubsub',
                   os.path.join(pkg_root, 'pubsub', 'commands'),
                   component='gcloud')
  loader.AddModule('services',
                   os.path.join(pkg_root, 'service_management', 'subcommands'))
  loader.AddModule('source',
                   os.path.join(pkg_root, 'source', 'commands'))
  loader.AddModule('sql', os.path.join(pkg_root, 'sql', 'tools'),
                   component='sql')
  loader.AddModule('test', os.path.join(pkg_root, 'test', 'commands'))

  # Check for updates on shutdown but not for any of the updater commands.
  loader.RegisterPostRunHook(UpdateCheck,
                             exclude_commands=r'gcloud\.components\..*')
  return loader.Generate()

_cli = CreateCLI()


def main():
  metrics.Started(START_TIME)
  # TODO(user): Put a real version number here
  metrics.Executions(
      'gcloud',
      local_state.InstallationState.VersionForInstalledComponent('core'))
  try:
    _cli.Execute()
  except Exception as err:  # pylint:disable=broad-except
    if isinstance(err, KeyboardInterrupt):
      raise
    traceback.print_exc()
    sys.exit(1)

if __name__ == '__main__':
  try:
    main()
  except KeyboardInterrupt:
    CTRLCHandler(None, None)
